src/main/java/bio/terra/workspace/service/datareference/model/ReferenceObject.java

```diff
@@ -0,0 +1,32 @@
+package bio.terra.workspace.service.datareference.model;
+
+import bio.terra.workspace.service.datareference.exception.InvalidDataReferenceException;
+import java.util.Map;
+
+/**
+ * An interface representing the subject of a data reference.
+ *
+ * <p>All referenced objects are read-only value objects described by a set of string key-value
+ * pairs. The set of keys available are specific to each referenced object type, as are validation
+ * rules. ReferenceObject implementations must be fully serializable by the output of getProperties.
+ */
+public interface ReferenceObject {
+  Map<String, String> getProperties();
```

Comment:
I don't really understand what getProperties is for. Is this for serialization only? Or is there something inherent about ReferenceObjects that they are maps of strings to strings?
If it's only for serialization in FlightMaps or Databases, we can accomplish that in other ways, e.g. toJson & fromJson abstract methods.

This is kind of a funny interface in that there's no behavior on it that's about what ReferenceObjects are for. Not necessarily wrong, I'm just not sure what WM actually does with ReferenceObjects given this interface. I think it's serialization at the request & db layers, but there aren't obvious methods for that.

Reply:
I think all ReferenceObjects really are maps of strings to strings, but in practice all we really care about here is serializing them. Agreed that it's a bit weird as is - I think having serialize/deserialize methods instead of directly exposing a map makes it clearer what this is for.

That said, I think using the same serialize/deserialize methods for database storage introduces the risk of bugs where changing fields of `SnapshotReference` breaks existing values. Using separate serialization classes does add more boilerplate though, so I'm not sure if this solution is overkill or not.

Reply:
Missed this earlier, sorry. I think for both DAO and Stairway serialization, we need to support backwards compatibility and should have unit tests for the same - e.g. hard code the string value of all current/past versions and check they can be deserialized. So, as in other review comments, I'm not totally convinced we need 2 serialization formats for ReferenceObjects.